### PR TITLE
[Feature][MRV2][310P] Add temperature / top-k / top-p sampling for MRV2 on the 310P

### PR DESCRIPTION
**Pull request by Thiagor2002**
### What this PR does / why we need it?

Enable **temperature / top-k / top-p** sampling on Ascend **310P Model Runner V2**.

- First-version `Ascend310PSampler` only supported greedy (`argmax`) and rejected non-zero temperature.
- Mainline MRV2 uses Triton Gumbel sampling; 310P has no Triton, and NPU `exponential_` / large RNG can hang under ACLGraph (same issue fixed in MRV1).
- This PR reuses the **MRV1 inverse-CDF path** (`_random_sample_310p`: CPU uniform per request → NPU `softmax` + `cumsum` + `searchsorted`), while keeping the MRV2 `sampling_states` surface required by MTP draft `propose()`.
- Still rejects min_p / penalties / logprobs / logit processors (follow-up).

Refs: MRV1 post-sampling fixes (#8017 / #8077), CDF optimization (#12966).

RFC: #15577 

### Does this PR introduce _any_ user-facing change?

No. On 310P with `VLLM_USE_V2_MODEL_RUNNER=1`, requests may now use `temperature`, `top_k`, and `top_p` (previously raised `NotImplementedError`). Default greedy behavior is unchanged.

### How was this patch tested?

- UT: `tests/ut/_310p/test_model_runner_v2_310p.py` (temperature accept / scale / greedy; still rejects penalties)

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Review by linfeng-yuan — COMMENTED** [review]
Please add e2e test for sampling.

**Comment by Thiagor2002** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by Tflowers-0129 — APPROVED** [review]

**Comment by Thiagor2002** [comment]
> Please add e2e test for sampling.

i have added e2e test in the newest commit.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 1]
```diff
@@ -2,19 +2,54 @@
...
+    temperature: torch.Tensor,
+) -> None:
+    """In-place temperature scaling (Triton-free; skips temp in {0, 1})."""
+    temps = temperature[expanded_idx_mapping].to(dtype=torch.float32)
+    need_scale = (temps != 0.0) & (temps != 1.0)
+    if not bool(need_scale.any().item()):
+        return
+    # Work in FP32 for numerical stability, write back to logits dtype.
+    scaled = logits.to(torch.float32) / temps.unsqueeze(-1)
+    logits.copy_(torch.where(need_scale.unsqueeze(-1), scaled.to(logits.dtype), logits))
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:44]: Thanks. Not adopting this. Rows with `temp ∈ {0, 1}` are excluded via `need_scale` and never written back by `torch.where`; any intermediate Inf on those rows is discarded. This matches our MRV1-aligned path and has been validated on 310P — no change planned.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 2]
```diff
@@ -24,21 +59,39 @@ def __init__(
...
+
         temperature_gpu = torch.zeros(max_num_reqs, dtype=torch.float32, device=device)
         seeds_gpu = torch.zeros(max_num_reqs, dtype=torch.int64, device=device)
+        top_k_gpu = torch.full(
+            (max_num_reqs,),
+            fill_value=max(self.vocab_size, 1),
+            dtype=torch.int32,
+            device=device,
+        )
+        top_p_gpu = torch.ones(max_num_reqs, dtype=torch.float32, device=device)
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:81]: Thanks. Not adopting. Production init always passes a known `vocab_size`; treating the default as `max(vocab_size, 1)` is intentional and keeps top-k “disabled” semantics simple. Introducing a `-1` sentinel here is unnecessary for our serving path.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 3]
```diff
@@ -24,21 +59,39 @@ def __init__(
...
         self.sampling_states = SimpleNamespace(
             temperature=SimpleNamespace(gpu=temperature_gpu),
             seeds=SimpleNamespace(gpu=seeds_gpu),
+            top_k=SimpleNamespace(gpu=top_k_gpu),
+            top_p=SimpleNamespace(gpu=top_p_gpu),
+            vocab_size=self.vocab_size,
         )
+        self._temperature_np = np.zeros(max_num_reqs, dtype=np.float32)
+        self._top_k_np = np.full(max_num_reqs, fill_value=max(self.vocab_size, 1), dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:90]: keeping the current default `(max(vocab_size, 1) / 1` when unknown). Not adopting the `-1` sentinel change.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 4]
```diff
@@ -69,21 +116,134 @@ def add_request(
...
+        temperature = float(sampling_params.temperature)
+        top_p = float(sampling_params.top_p)
+        top_k = int(sampling_params.top_k)
+        if self.vocab_size <= 0:
+            # Late-bind from logits on first sample if caller omitted vocab_size.
+            vocab_size = max(top_k, 1)
+        else:
+            vocab_size = self.vocab_size
+        if top_k <= 0 or top_k > vocab_size:
+            top_k = vocab_size
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:136]: Not adopting. We resolve disabled`/<=0` top-k`to vocab size once vocab is known; the suggested `-1` keep path is not needed for the MRV2 310P MTP serve flow and would diverge from the current MRV1-aligned handling.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 5]
```diff
@@ -69,21 +116,134 @@ def add_request(
...
 
+    def _maybe_bind_vocab_size(self, vocab_size: int) -> None:
+        if self.vocab_size > 0 or vocab_size <= 0:
+            return
+        self.vocab_size = int(vocab_size)
+        self.sampling_states.vocab_size = self.vocab_size
+        # Replace unset top_k defaults (initialized to 1 when vocab was unknown).
+        unset = self._top_k_np <= 1
+        self._top_k_np[unset] = self.vocab_size
+        self.sampling_states.top_k.gpu[unset] = self.vocab_size
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:170]: Not adopting. Late-bind is only a fallback when vocab was omitted at construction; in our real init path vocab is set upfront, so the suggested sentinel logic isn’t required. Leaving the current `<= 1` replace as-is.

File: tests/ut/_310p/test_model_runner_v2_310p.py [hunk 6]
```diff
@@ -382,11 +382,41 @@ def test_copy_kv_cache_blocks_flattens_mamba_lists() -> None:
...
+    sampler.add_request(0, 2, SamplingParams(temperature=0))
+    logits = torch.tensor([[0.1, 3.0, 0.2, 0.0]], dtype=torch.float32)
+    input_batch = SimpleNamespace(
+        expanded_idx_mapping=torch.tensor([0], dtype=torch.int32),
+        idx_mapping_np=np.array([0], dtype=np.int32),
+        num_reqs=1,
+        seq_lens=torch.ones(1, dtype=torch.int32),
+    )
+    out = sampler(logits, input_batch)
+    assert int(out.sampled_token_ids.view(-1)[0].item()) == 1
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply tests/ut/_310p/test_model_runner_v2_310p.py:419]: Thanks. Not adopting for now. Existing UT already covers temperature / greedy / unsupported params for the production path. We’ll skip the extra late-bind/`top_k=1` case since that construction path isn’t used in our serve setup.

File: vllm_ascend/_310p/worker/v2/sampler.py [hunk 7]
```diff
@@ -2,19 +2,54 @@
 # Copyright (c) 2026 Huawei Technologies Co., Ltd. All Rights Reserved.
 # mypy: ignore-errors
 
+"""Triton-free MRV2 sampler for Ascend 310P.
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/sampler.py:5]: Thanks for the suggestion. We’ll keep the current [Feature][MRV2][310P] title and PR description as-is — they already match our 310P MRV2 convention and cover the intent of this change. Not adopting the suggested rewrite.
